core/common/src/main/java/alluxio/collections/IndexedSet.java

```diff
@@ -97,14 +95,13 @@
  */
 @ThreadSafe
 public class IndexedSet<T> implements Iterable<T> {
-  /** All objects in the set. */
-  private final Set<T> mObjects = new HashSet<T>();
-  /** Map from field index to an index of field related object in the internal lists. */
-  private final Map<FieldIndex<T>, Integer> mIndexMap;
-  /** List of maps from value of a specific field to set of objects with the field value. */
-  private final List<Map<Object, Set<T>>> mSetIndexedByFieldValue;
-  /** Final object for synchronization. */
-  private final Object mLock = new Object();
+  /** All objects in the set. This set is required to guarantee uniqueness of objects. */
+  private final ConcurrentHashSet<T> mObjects = new ConcurrentHashSet<>(1024, 0.75f, 32);
```

Comment:
Hrmmm, using only indexes is an interesting point. Currently, we do not do any enforcement for a "primary" index, so this set enforces that. However, it seems possible to delegate that to the first index.

Also, iterator might be a bit more complicated, since the `IndexedSet` iterator would need an iterator of iterators. That sounds a bit complicated, especially when each element allows concurrent access.

I will have to investigate this approach.